polynote-spark/src/main/scala/polynote/kernel/remote/RemoteSparkKernel.scala

```diff
@@ -0,0 +1,322 @@
+    case None => IO(logger.error(s"Received response $rep, but there is no corresponding request outstanding"))
+    case Some(deferred) => deferred.complete(rep)
+  }
+
+  private def handleStreamStart(reqId: Int) = Option(streamRequests.get(reqId)) match {
+    case None => IO(logger.error(s"Received stream start for $reqId, but there is no corresponding stream request outstanding"))
+    case Some(deferred) => Queue.unbounded[IO, Option[Result]] >>= deferred.complete
+  }
+
+  private def withQueue(reqId: Int, fn: Queue[IO, Option[Result]] => IO[Unit]) = Option(streamRequests.get(reqId)) match {
+    case None => IO(logger.error(s"Received stream start for $reqId, but there is no corresponding stream request outstanding"))
+    case Some(deferred) => Queue.unbounded[IO, Option[Result]] >>= fn
+  }
+
+  private def transformResult(result: Result): Result = result match {
+    case rv @ ResultValue(_, _, reprs, _, _, _) =>
+      // the ResultValue will be missing its JVM value and its type. Also, any handle-based reprs won't refer to handles
+      // within this JVM. We shouldn't need the value or type outside of the actual kernel, but we do have to fix
+      // the repr handles.
+      rv.copy(reprs = reprs.map {
+        case s @ StreamingDataRepr(_, _, _) => StreamingDataRepr.fromHandle(transformRepr(s))
+        case l @ LazyDataRepr(_, _) => LazyDataRepr.fromHandle(transformRepr(l))
+        case u @ UpdatingDataRepr(remoteHandle, dataType) =>
+          val local = UpdatingDataRepr(dataType)
+          updatingHandleMapping.put(remoteHandle, local.handle)
+          local
+        case repr => repr
+      })
+  }
+
+  private def consumeResponses(responses: Stream[IO, RemoteResponse]): Stream[IO, Unit] = responses.flatMap {
+    case Announce(remoteAddress) => Stream.eval(parseHostPort(remoteAddress)).evalMap(addr => remoteAddr.complete(addr).handleErrorWith(_ => IO.unit))
+    case KernelStatusResponse(update) => Stream.eval(statusUpdates.publish1(update))
+    case StreamStarted(reqId) => Stream.eval(handleStreamStart(reqId))
+    case StreamEnded(reqId) => Stream.eval(withQueue(reqId, _.enqueue1(None)))
+    case ResultStreamElements(reqId, elements) => Stream.eval(withQueue(reqId, q => Stream.emits(elements).map(Some(_)).through(q.enqueue).compile.drain))
+    case rep: RequestResponse => Stream.eval(handleOneResponse(rep))
+  }
+
+  /**
+    * Launches the spark-submit process which will run the RemoteSparkKernelClient
+    */
+  def init: IO[Unit] = for {
+    _            <- statusUpdates.publish1(taskInfo("Starting kernel process"))
+    notebook     <- getNotebook()
+    connectFiber <- IO(server.accept()).start
+    sparkConfig   = config.spark ++ notebook.config.flatMap(_.sparkConfig).getOrElse(Map.empty)
+    sparkArgs     = sparkConfig.flatMap(kv => Seq("--conf", s"${kv._1}=${kv._2}"))
+    command       = Seq("spark-submit", "--class", classOf[RemoteSparkKernelClient].getName) ++ sparkArgs ++ Seq(jarURL, serverHostPort)
```

Comment:
Hmm, looks like the jar URL isn't used by the `RemoteSparkKernelClient`? Also, I don't see where `"--remoteAddress"` is set but it looks like that's necessary?

Reply:
Yeah, when I went to add that arg I realized I should stop and factor out the comms and deployment stuff. But you're right, it is necessary.

Reply:
The JAR URL is used by `spark-submit` - it's the main JAR to use (in cluster mode you have to use a URL that's accessible from the remote driver)

Reply:
Makes sense!